### PR TITLE
Extract Error Handling Paths in TransportService

### PR DESCRIPTION
**Pull request by original-brownbear**
The error handling in the internal send method is very complex and more complex
than the normal code path of that method. Extracting it here to make this (very hot)
method easier to read and compile better/smaller. Also, extract the error handling
paths for the other sending methods to deduplicate the code a little and shorten
those methods as well.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/transport/TransportService.java [hunk 1]
```diff
@@ -572,41 +572,28 @@ public HandshakeResponse(StreamInput in) throws IOException {
...
+            clusterName = new ClusterName(in);
+        }
 
+        private void failIfIncompatibleBuild(@Nullable Exception cause) {
             if (isIncompatibleBuild(version, buildHash)) {
                 throw new IllegalArgumentException(
-                    "remote node ["
-                        + discoveryNode
-                        + "] is build ["
+                    "unidentifiable remote node is build ["
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/TransportService.java:585]: I think we need to keep the `remote node [...]` bit in cases where the node isn't unidentifiable. That's almost always the path we use and without that it's almost impossible to determine which node is the problem.

File: server/src/main/java/org/elasticsearch/transport/TransportService.java [hunk 2]
```diff
@@ -764,17 +751,26 @@ public String toString() {
...
+    private <T extends TransportResponse> void handleSendRequestException(
+        Transport.Connection connection,
+        String action,
+        TransportResponseHandler<T> handler,
+        Exception ex
+    ) {
+        // the caller might not handle this so we invoke the handler
+        final TransportException te;
+        if (ex instanceof TransportException) {
+            te = (TransportException) ex;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/TransportService.java:767]: Avoid cast?

```suggestion
        if (ex instanceof TransportException tex) {
            te = tex;
```

File: server/src/main/java/org/elasticsearch/transport/TransportService.java [hunk 3]
```diff
@@ -974,12 +965,16 @@ public void onAfter() {
             }
 
         } catch (Exception e) {
-            try {
-                channel.sendResponse(e);
-            } catch (Exception inner) {
-                inner.addSuppressed(e);
-                logger.warn(() -> new ParameterizedMessage("failed to notify channel of error message for action [{}]", action), inner);
-            }
+            handleSendToLocalException(channel, e, action);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/TransportService.java:968]: Is this ever expected? If not, we should `assert false` here. If we do expect it sometimes I think we should fork to the appropriate executor too.

**Review by DaveCTurner — COMMENTED** [review]
Seems sensible, cleaning this up is long overdue. I left one blocking comment and a couple of nits.

File: server/src/main/java/org/elasticsearch/transport/TransportService.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/TransportService.java:977]: This was handling the exceptions from the `SAME` path which now that you ask is really confusing. I added a dedicated try-catch for that now and an assert here.

File: server/src/main/java/org/elasticsearch/transport/TransportService.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/transport/TransportService.java]: ++

File: server/src/main/java/org/elasticsearch/transport/TransportService.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/transport/TransportService.java:977]: Oh right so it was. Yes I prefer this now.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**